php_phongo.c

```diff
@@ -2550,6 +2610,11 @@ static char* php_phongo_manager_make_client_hash(const char* uri_string, zval* o
 		hash      = pestrndup(ZSTR_VAL(var_buf.s), *hash_len, 1);
 	}
 
+	if (serializable_driver_options && free_driver_options) {
+		zval_ptr_dtor(serializable_driver_options);
```

Comment:
I think this may be incorrect for PHP 5.x, as `zval_ptr_dtor()` expects a `zval**` even though it immediately dereferences it to a `zval*` when calling `i_zval_ptr_dtor()` (see: [here](http://lxr.jpauli.tech/source/xref/php-5.6/Zend/zend_execute.c#96)).

That said, I'm curious if we even need to do this since a non-null `serializable_driver_options` means we've added it to `args`, which itself is going to be passed to `zval_ptr_dtor()` later on. As-is, this may be a double-dtor (I'm not sure how that compares to a double-free).

Of course, we still need to call `efree()` to clean up the zval allocation within `php_phongo_manager_prepare_manager_for_hash`, but I think that may be something to do _after_ the call to `zval_ptr_dtor(&args);`. Does that make sense?

It may be helpful to look at `php_phongo_prep_legacy_option()` and `php_phongo_prep_legacy_option_free()` for prior art, since they do something similar.

Reply:
> I think this may be incorrect for PHP 5.x, as `zval_ptr_dtor()` expects a `zval**` even though it immediately dereferences it to a `zval*` when calling `i_zval_ptr_dtor()` (see: [here](http://lxr.jpauli.tech/source/xref/php-5.6/Zend/zend_execute.c#96)).

The code in question is in a conditional for PHP 7. The following code block that covers PHP 5 passes the expected `zval**`:
```
	if (serializable_driver_options && free_driver_options) {
		zval_ptr_dtor(&serializable_driver_options);
	}
```

> That said, I'm curious if we even need to do this since a non-null `serializable_driver_options` means we've added it to `args`, which itself is going to be passed to `zval_ptr_dtor()` later on. As-is, this may be a double-dtor (I'm not sure how that compares to a double-free).
> 
> Of course, we still need to call `efree()` to clean up the zval allocation within `php_phongo_manager_prepare_manager_for_hash`, but I think that may be something to do _after_ the call to `zval_ptr_dtor(&args);`. Does that make sense?

Great catch, that's absolutely correct. If we duplicate the driver options, the new object will be freed when `zval_ptr_dtor(&args)` recurses into nested zvals. I didn't quite make the entire connection, but I did add a `Z_ADDREF_P` call for the case where we don't need to duplicate the driverOptions and append the original zval to `args`. I've changed the code to no longer call `zval_ptr_dtor(&serializable_driver_options)`, but only `efree` it after calling `zval_ptr_dtor(&args)`. For PHP 5, this means we don't need to clean up anything since we only called `zval_ptr_dtor` in the first place.

> It may be helpful to look at `php_phongo_prep_legacy_option()` and `php_phongo_prep_legacy_option_free()` for prior art, since they do something similar.

I originally looked at that implementation, but misinterpreted the handling of `options` vs. `args` in these two methods. Your comment helped clear up that confusion.